FP on S1118 with Lombok constructors

bannmann opened the pull request:
This fixes false positives of S1118 ("Utility classes should not have public constructors") regarding lombok-generated constructors; see my post at https://groups.google.com/forum/#!topic/sonarqube/hu2sITqPyQs

- [x] Use the following formatting style: [SonarSource/sonar-developer-toolset](https://github.com/SonarSource/sonar-developer-toolset#code-style)
- [x] Unit tests are passing and you provided a unit test for your fix
- [ ] ITs should pass : To run ITs locally, checkout the [README](https://github.com/SonarSource/sonar-java/blob/master/README.md) of the project.
    - **_My local ruling test failed, but had exactly the same results as with `master`._**
- [ ] If there is a [Jira](http://jira.sonarsource.com/browse/SONARJAVA) ticket available, please make your commits and pull request start with the ticket number (SONARJAVA-XXXX)
    - **_There was no reaction to my post, so there is no Jira ticket._**

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/filters/LombokFilter.java:
```diff
@@ -116,4 +121,42 @@ private static boolean usesAnnotation(ClassTree classTree, List<String> annotati
+  }
+
+  private static boolean generatesPrivateAccess(SymbolMetadata metadata, String annotation) {
+    for (SymbolMetadata.AnnotationValue annotationValue : metadata.valuesForAnnotation(annotation)) {
```
You can use a Stream directly here:
```
return metadata.valuesForAnnotation(annotation).anyMatch(av -> ... );
```

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/filters/LombokFilter.java:
```diff
@@ -116,4 +121,42 @@ private static boolean usesAnnotation(ClassTree classTree, List<String> annotati
+
+  private static boolean generatesPrivateAccess(SymbolMetadata metadata, String annotation) {
+    for (SymbolMetadata.AnnotationValue annotationValue : metadata.valuesForAnnotation(annotation)) {
+      if (Objects.equals(annotationValue.name(), "access") && Objects.equals(getAccessLevelValue(annotationValue.value()), "PRIVATE")) {
```
I think we could simplify this a bit by removing usage of `Objects`:
```
if ("access".equals(annotationValue.name())) && "PRIVATE".equals(getAccessLevelValue(annotationValue.value())) {
```

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/filters/LombokFilter.java:
```diff
@@ -116,4 +121,42 @@ private static boolean usesAnnotation(ClassTree classTree, List<String> annotati
+      return true;
+    }
+    SymbolMetadata metadata = classTree.symbol().metadata();
+    for (String annotation : GENERATE_CONSTRUCTOR) {
```
You can replace this part by :
```
return GENERATE_CONSTRUCTOR.stream().anyMatch(metadata -> metadata.isAnnotatedWith(annotation) && generatesPrivateAccess(metadata, annotation));
```

I would also move the `isAnnotatedWith(annotation)`directly in the `gernatePrivateAccess` method, so it makes the stream simpler.

m-g-sonar commented on java-checks/src/main/java/org/sonar/java/filters/LombokFilter.java:
```diff
@@ -116,4 +121,42 @@ private static boolean usesAnnotation(ClassTree classTree, List<String> annotati
+
+  @Nullable
+  private static String getAccessLevelValue(Object value) {
+    String retentionValue = null;
```
This is not a `retentionValue`... that's the Access level. Rename the variable or simply drop it to return immediately the value when getting it.

`RetentionValue` is variable name from `java.lang.annotation.Retention` (copy-pasted from ReflectionOnNonRuntimeAnnotationCheck` ?).

m-g-sonar commented:
Thanks for the update @bannmann. Now, everything looks good to me regarding your contribution, good job!

It's perfect to me to keep commits separated during review phase. It's indeed when opening a PR that we usually appreciate to have everything squashed together (if it makes sense).

In a few minutes I'm going to open a new PR with your changes, which will trigger our internal QA. If everything goes green, I'll then close mine and merge your PR. I'll let you know if there is anything to change according to QA (no ideas why ITs were not working on your machine).

Cheers,
Michael

m-g-sonar commented:
Okay, 2 issues spotted by SonarQube analysis: https://next.sonarqube.com/sonarqube/project/issues?id=org.sonarsource.java%3Ajava&pullRequest=2052&resolved=false

Sorry I missed that. Anyway, from my point of view, the NPE can not occur, because you are explicitly testing if `isAnnotatedWith(...)` before trying to access annotations values... Nevertheless, it made me realize we can be a bit more efficient/explicit.

Because `valuesForAnnotation()` is actually returning null if not annotated, I would rewrite the following
```java
private static boolean generatesPrivateAccess(SymbolMetadata metadata, String annotation) {
  return metadata.isAnnotatedWith(annotation) && metadata.valuesForAnnotation(annotation).stream().anyMatch(av -> "access".equals(av.name()) && "PRIVATE".equals(getAccessLevelValue(av.value())));
}
```
to :
```java
private static boolean generatesPrivateAccess(SymbolMetadata metadata, String annotation) {
  List<AnnotationValue> values = metadata.valuesForAnnotation(annotation);
  return values != null && values.stream().anyMatch(av -> "access".equals(av.name()) && "PRIVATE".equals(getAccessLevelValue(av.value())));
}
```

And now,  we could simplify a bit more by also changing the caller method to:
```java
private static boolean generatesPrivateConstructor(ClassTree classTree) {
  if (usesAnnotation(classTree, GENERATE_PRIVATE_CONSTRUCTOR)) {
    return true;
  }
  SymbolMetadata metadata = classTree.symbol().metadata();
  return GENERATE_CONSTRUCTOR.stream()
    .map(metadata::valuesForAnnotation)
    .filter(Objects::nonNull)
    .anyMatch(LombokFilter::generatesPrivateAccess);
}

private static boolean generatesPrivateAccess(List<AnnotationValue> values) {
  return values.stream().anyMatch(av -> "access".equals(av.name()) && "PRIVATE".equals(getAccessLevelValue(av.value())));
}
```

I let you choose which approach you prefer, between the null check or the full rewriting, and I'll update the duplicate PR once updated.

Cheers,
Michael

bannmann commented:
I went with the rewrite. The null safety created by the `isAnnotatedWith()` check is way too subtle for my taste, too, and keeping it plus adding a redundant null check to satisfy SonarQube is just clumsy.

By the way, I also played with using `flatMap()` instead of `map()` so that the `generatesPrivateAccess()` would not need to operate on `List`, but in the end that did make the code clearer or more concise in any way.

Added commit and rebased on new master.

m-g-sonar commented:
Hey @bannmann ,

Thanks for the update and for your contribution. Merged into master!

The fix will be part of release 5.5 of SonarJava, to be delivered in 2-3 weeks.

Cheers,
Michael
